src/packet/packetlist.js

```diff
@@ -70,65 +69,121 @@ class PacketList extends Array {
     if (config.additionalAllowedPackets.length) {
       allowedPackets = { ...allowedPackets, ...util.constructAllowedPackets(config.additionalAllowedPackets) };
     }
+    const tagsRead = [];
     this.stream = streamTransformPair(bytes, async (readable, writable) => {
+      const reader = streamGetReader(readable);
       const writer = streamGetWriter(writable);
-      const writtenTags = [];
       try {
+        let useStreamType = util.isStream(readable);
         while (true) {
           await writer.ready;
-          const done = await readPackets(readable, async parsed => {
+          let error;
+          let wasStream;
+          await readPacket(reader, useStreamType, async parsed => {
             try {
               if (parsed.tag === enums.packet.marker || parsed.tag === enums.packet.trust || parsed.tag === enums.packet.padding) {
-                // According to the spec, these packet types should be ignored and not cause parsing errors, even if not esplicitly allowed:
+                // According to the spec, these packet types should be ignored and not cause parsing errors, even if not explicitly allowed:
                 // - Marker packets MUST be ignored when received: https://github.com/openpgpjs/openpgpjs/issues/1145
                 // - Trust packets SHOULD be ignored outside of keyrings (unsupported): https://datatracker.ietf.org/doc/html/rfc4880#section-5.10
                 // - [Padding Packets] MUST be ignored when received: https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-crypto-refresh#name-padding-packet-tag-21
                 return;
               }
               const packet = newPacketFromTag(parsed.tag, allowedPackets);
+              // Unknown packets throw in the call above, we ignore them
+              // in the grammar checker.
+              tagsRead.push(parsed.tag);
               packet.packets = new PacketList();
               packet.fromStream = util.isStream(parsed.packet);
-              await packet.read(parsed.packet, config);
+              wasStream = packet.fromStream;
+              try {
+                await packet.read(parsed.packet, config);
+              } catch (e) {
+                if (!(e instanceof UnsupportedError)) {
+                  e.name = 'MalformedPacketError';
+                }
+                throw e;
+              }
               await writer.write(packet);
-              writtenTags.push(parsed.tag);
-              // The `writtenTags` are only sensitive if we are parsing an _unauthenticated_ decrypted stream,
-              // since they can enable an decryption oracle.
-              // It's responsibility of the caller to pass a `grammarValidator` that takes care of
-              // postponing error reporting until the data has been authenticated.
-              grammarValidator?.(writtenTags, true, config);
```

Comment:
I still think we should not go ahead with parsing data that we know to be invalid

Reply:
Conceptually, the grammar checker operates on a packet sequence, i.e. a list of already-parsed packets. Streaming makes it possible to stop parsing earlier but I don't think it's necessary. As long as we don't use an invalid packet sequence higher up the stack, I don't think there's any risk here of parsing packets in a sequence that turns out to be invalid.

Reply:
The reason why `allowedPackets` was implemented this way is actually because of tree shaking: if you only import `openpgp.verify`, then it only imports signature packets and literal/compressed data packets, and so parsing an encrypted data packet is impossible.

What's the security risk of parsing a packet that's not allowed, if we throw before using it?